File: app/src/main/java/com/gystry/pjetpack/AbsViewModel.java
package com.gystry.pjetpack;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;
import androidx.paging.DataSource;
import androidx.paging.LivePagedListBuilder;
import androidx.paging.PagedList;

/**
 * @author gystry
 * 创建日期：2020/7/9 11
 * 邮箱：gystry@163.com
 * 描述：
 */
public abstract class AbsViewModel<T> extends ViewModel {

    private LiveData<PagedList<T>> pageData;
    private DataSource dataSource;
    private MutableLiveData<Boolean> boundaryPageData=new MutableLiveData<>();

    public AbsViewModel() {
        PagedList.Config config = new PagedList.Config.Builder()
                .setPageSize(10)
                .setInitialLoadSizeHint(10)//第一次加载数据的时候（初始化加载）加载的数据数量
//                .setMaxSize(100)  最大数值
//                .setEnablePlaceholders(false)  占位符，当要展示100条数据的时候，但是实际数据只有十条，那到第十一条的时候就会用占位符展示
//                .setPrefetchDistance()//在距离列表底部还剩几个数据的时候就开始加载下一页，如果不设置，就默认是一页的数量
                .build();

        pageData = new LivePagedListBuilder(factory, config)
                .setInitialLoadKey(0)
//                .setFetchExecutor()
                .setBoundaryCallback(boundaryCallback)//监听pagelist数据加载的状态
                .build();
    }

    public LiveData<PagedList<T>> getPageData() {
        return pageData;
    }

    public DataSource getDataSource() {
        return dataSource;
    }

    public MutableLiveData<Boolean> getBoundaryPageData() {
        return boundaryPageData;
    }

    PagedList.BoundaryCallback<T> boundaryCallback = new PagedList.BoundaryCallback<T>() {
        @Override
        public void onZeroItemsLoaded() {
            //从PagedList的数据源的初始加载返回零项时被调用
            boundaryPageData.postValue(false);
        }

        @Override
        public void onItemAtFrontLoaded(@NonNull T itemAtFront) {
            //当列表的第一条数据被加载的时候会被调用
            boundaryPageData.postValue(true);
        }

        @Override
        public void onItemAtEndLoaded(@NonNull T itemAtEnd) {
            //当pageList的最后一条数据被加载到屏幕上的时候，这个方法会被调用
            super.onItemAtEndLoaded(itemAtEnd);
        }
    };

    DataSource.Factory factory = new DataSource.Factory() {
        @NonNull
        @Override
        public DataSource create() {
            dataSource = createDataSource();
            return dataSource;
        }
    };

    public abstract DataSource createDataSource();
}

File: appkt/src/main/java/com/gystry/appkt/AbsViewModel.kt
package com.gystry.appkt

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import androidx.paging.DataSource
import androidx.paging.LivePagedListBuilder
import androidx.paging.PagedList
import com.gystry.appkt.model.Feed

/**
 * @author gystry
 * 创建日期：2021/4/1 17
 * 邮箱：gystry@163.com
 * 描述：
 */
abstract class AbsViewModel<T> : ViewModel() {
    var createDataSource: DataSource<Int, T>? = null
    val boundaryPageData
        get() = MutableLiveData<Boolean>()
    val cacheDataSource
        get() = MutableLiveData<PagedList<Feed>>()
    protected val config: PagedList.Config by lazy {
        PagedList.Config.Builder()
                .setPageSize(10)
                .setInitialLoadSizeHint(12)
                .build()
    }

    //监听pagelist数据加载的状态
    private val boundaryCallback: PagedList.BoundaryCallback<T> by lazy {
        object : PagedList.BoundaryCallback<T>() {
            override fun onZeroItemsLoaded() {
                //从PagedList的数据源的初始加载返回零项时被调用
                boundaryPageData.postValue(false)
            }

            override fun onItemAtFrontLoaded(itemAtFront: T) {
                //当列表的第一条数据被加载的时候会被调用
                boundaryPageData.postValue(true)
            }

            override fun onItemAtEndLoaded(itemAtEnd: T) {
                //当pageList的最后一条数据被加载到屏幕上的时候，这个方法会被调用
                super.onItemAtEndLoaded(itemAtEnd)
            }
        }
    }

    private val factory: DataSource.Factory<Int, T> by lazy {
        object : DataSource.Factory<Int, T>() {
            override fun create(): DataSource<Int, T> {
                createDataSource = createDataSource()
                return createDataSource as DataSource<Int, T>
            }
        }
    }

    fun getDataSource():DataSource<Int, T>?{
        return createDataSource
    }

    val pageData: LiveData<PagedList<T>>
        get() = LivePagedListBuilder(factory, config)
                .setInitialLoadKey(0)//加载初始化数据的时候需要传递的数据
                .setBoundaryCallback(boundaryCallback)//
                .build()


    abstract fun createDataSource(): DataSource<Int, T>
}